lib/async/redis/context/pipeline.rb

```diff
@@ -0,0 +1,44 @@
+module Async
+	module Redis
+		module Context
+
+			# This class accumulates commands and sends several of them in a single
+			# request, instead of sending them one by one.
+			class Pipeline
+				include Methods::Strings
+				include Methods::Keys
+				include Methods::Lists
+
+				def initialize(connection_pool)
+					@pool = connection_pool
+					@connection = connection_pool.acquire
+
+					# Each command is an array where the first element is the name of the
+					# command ('SET', 'GET', etc.) and the rest of elements are the
+					# parameters for that command.
+					# Ex: ['SET', 'some_key', 42].
+					@commands = []
+				end
+
+				# This method just accumulates the commands and their params.
+				def call(*args)
+					@commands << args
```

Comment:
I'm not familiar with all the classes in `Async::IO`. Let me take a look at this and I'll give you an aswer a bit later.

Reply:
So if you exposed `@stream#flush` as `RESP#flush` class, what you could do is instead of accumulating the commands in an array, you could write them straight to the stream buffer and then just flush the buffer from `Pipeline#run`.

Reply:
With that you could get rid of the whole `write_pipeline` method in `RESP`.

Reply:
> These responses should be held somewhere along redis' internal buffer -> os buffers -> async-io stream's buffer until we read them out.

`Async::IO::Stream` does it's best to disable as much buffering as possible in the OS/network stack. See `Async::IO::Stream#sync=`

Reply:
I'm really bad at describing what I mean with human words, here is some code:
https://github.com/socketry/async-redis/blob/add-pipelines-support/lib/async/redis/context/pipeline.rb

Reply:
Here is the spec https://github.com/socketry/async-redis/blob/add-pipelines-support/spec/async/redis/context/pipeline_spec.rb Of course the way it reads back the responses from the requests is very cumbersome, that's definitely something to be addressed.

Reply:
That makes sense. I guess it's okay, but I think we should also support nested async so users can actually do something with the responses in an async fashion. But that's additive to what you've proposed already.

Reply:
@huba I took a look at the `Async::IO::Stream` module and your changes look good to me.